Flaky ZeroDivisionError test

While working on #70 , I occasionally encountered the following error in pytest:

```text
2021-09-28T23:18:17.1539458Z Installing collected packages: pyparsing, toml, py, pluggy, packaging, iniconfig, attrs, pytest
2021-09-28T23:18:17.5669452Z Successfully installed attrs-21.2.0 iniconfig-1.1.1 packaging-21.0 pluggy-1.0.0 py-1.10.0 pyparsing-2.4.7 pytest-6.2.5 toml-0.10.2
2021-09-28T23:18:18.3053147Z ============================= test session starts ==============================
2021-09-28T23:18:18.3056239Z platform linux -- Python 3.6.15, pytest-6.2.5, py-1.10.0, pluggy-1.0.0
2021-09-28T23:18:18.3057466Z rootdir: /home/runner/work/taskgraph/taskgraph
2021-09-28T23:18:18.3058344Z collected 46 items
2021-09-28T23:18:18.3058894Z 
2021-09-28T23:18:24.6466033Z tests/test_task.py .F............................................        [100%]
2021-09-28T23:18:24.6466851Z 
2021-09-28T23:18:24.6467685Z =================================== FAILURES ===================================
2021-09-28T23:18:24.6468341Z _______________________ TaskGraphTests.test_broken_task ________________________
2021-09-28T23:18:24.6468740Z 
2021-09-28T23:18:24.6469367Z self = <test_task.TaskGraphTests testMethod=test_broken_task>
2021-09-28T23:18:24.6469827Z 
2021-09-28T23:18:24.6470200Z     def test_broken_task(self):
2021-09-28T23:18:24.6471606Z         """TaskGraph: Test that a task with an exception won't hang."""
2021-09-28T23:18:24.6472530Z         task_graph = taskgraph.TaskGraph(self.workspace_dir, 1)
2021-09-28T23:18:24.6473138Z         _ = task_graph.add_task(
2021-09-28T23:18:24.6473850Z             func=_div_by_zero, task_name='test_broken_task')
2021-09-28T23:18:24.6474341Z         task_graph.close()
2021-09-28T23:18:24.6474901Z         with self.assertRaises(ZeroDivisionError):
2021-09-28T23:18:24.6475468Z >           task_graph.join()
2021-09-28T23:18:24.6476015Z E           AssertionError: ZeroDivisionError not raised
2021-09-28T23:18:24.6476428Z 
2021-09-28T23:18:24.6476838Z tests/test_task.py:479: AssertionError
2021-09-28T23:18:24.6477844Z ------------------------------ Captured log call -------------------------------
2021-09-28T23:18:24.6478667Z DEBUG    taskgraph.Task:Task.py:265 /tmp/tmpkq0uigu5 already exists, no need to make it
2021-09-28T23:18:24.6479770Z DEBUG    taskgraph.Task:Task.py:1546 file:///tmp/tmpkq0uigu5/taskgraph_data.db?mode=ro exists: True
2021-09-28T23:18:24.6480601Z DEBUG    taskgraph.Task:Task.py:126 Starting logging worker
2021-09-28T23:18:24.6481426Z DEBUG    taskgraph.Task:Task.py:681 multithreaded: test_broken_task (0) sending to new task queue.
2021-09-28T23:18:24.6482615Z DEBUG    taskgraph.Task:Task.py:688 sending task test_broken_task (0) right away
2021-09-28T23:18:24.6483349Z DEBUG    taskgraph.Task:Task.py:804 Closing taskgraph.
2021-09-28T23:18:24.6484127Z DEBUG    taskgraph.Task:Task.py:1073 _call check if precalculated test_broken_task (0)
2021-09-28T23:18:24.6485193Z DEBUG    taskgraph.Task:Task.py:1160 file_stat_list: []
2021-09-28T23:18:24.6486007Z DEBUG    taskgraph.Task:Task.py:1161 other_arguments: [[], {}]
2021-09-28T23:18:24.6486790Z DEBUG    taskgraph.Task:Task.py:1546 file:///tmp/tmpkq0uigu5/taskgraph_data.db?mode=ro exists: True
2021-09-28T23:18:24.6487730Z DEBUG    taskgraph.Task:Task.py:1188 not precalculated, Task hash does not exist (test_broken_task (0))
2021-09-28T23:18:24.6488627Z DEBUG    taskgraph.Task:Task.py:1189 is_precalculated full task info: Task object 140694776114704:
2021-09-28T23:18:24.6489263Z 
2021-09-28T23:18:24.6489776Z {'exception_object': None,
2021-09-28T23:18:24.6490353Z  'ignore_directories': True,
2021-09-28T23:18:24.6490880Z  'ignore_path_list': [],
2021-09-28T23:18:24.6491363Z  'priority': 0,
2021-09-28T23:18:24.6491920Z  'self._reexecution_info': {'args_clean': [],
2021-09-28T23:18:24.6492525Z                             'file_stat_list': [],
2021-09-28T23:18:24.6493092Z                             'func_name': '_div_by_zero',
2021-09-28T23:18:24.6493638Z                             'kwargs_clean': {},
2021-09-28T23:18:24.6494218Z                             'other_arguments': [[], {}],
2021-09-28T23:18:24.6496811Z                             'source_code_hash': 'ae7a008fc3ac50d2882dd35ed5a9f31cdd73fcf9'},
2021-09-28T23:18:24.6497789Z  'self._result': None,
2021-09-28T23:18:24.6498720Z  'target_path_list': [],
2021-09-28T23:18:24.6499829Z  'task_id_hash': 'b31ea4a4abbc9d16ae2e9e7f2ddfb62ff1929271',
2021-09-28T23:18:24.6500661Z  'task_name': 'test_broken_task (0)',
2021-09-28T23:18:24.6501498Z  'task_reexecution_hash': '9310be23d42bb7e5e339e2ce11b843f64c7d8780'}
2021-09-28T23:18:24.6502349Z DEBUG    taskgraph.Task:Task.py:1077 not precalculated test_broken_task (0)
2021-09-28T23:18:24.6503104Z DEBUG    taskgraph.Task:Task.py:1084 apply_async for task test_broken_task (0)
2021-09-28T23:18:24.6504585Z ERROR    taskgraph.Task:Task.py:469 A taskgraph _task_executor failed on Task test_broken_task (0). Terminating taskgraph.
2021-09-28T23:18:24.6505616Z multiprocessing.pool.RemoteTraceback: 
2021-09-28T23:18:24.6506233Z """
2021-09-28T23:18:24.6506617Z Traceback (most recent call last):
2021-09-28T23:18:24.6508616Z   File "/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/multiprocessing/pool.py", line 119, in worker
2021-09-28T23:18:24.6509568Z     result = (True, func(*args, **kwds))
2021-09-28T23:18:24.6510269Z   File "/home/runner/work/taskgraph/taskgraph/tests/test_task.py", line 91, in _div_by_zero
2021-09-28T23:18:24.6510842Z     return 1/0
2021-09-28T23:18:24.6511273Z ZeroDivisionError: division by zero
2021-09-28T23:18:24.6511693Z """
2021-09-28T23:18:24.6511891Z 
2021-09-28T23:18:24.6512610Z The above exception was the direct cause of the following exception:
2021-09-28T23:18:24.6513053Z 
2021-09-28T23:18:24.6513420Z Traceback (most recent call last):
2021-09-28T23:18:24.6514869Z   File "/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/taskgraph-0.0.post1+g523403a-py3.6.egg/taskgraph/Task.py", line 462, in _task_executor
2021-09-28T23:18:24.6515838Z     task._call()
2021-09-28T23:18:24.6517223Z   File "/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/taskgraph-0.0.post1+g523403a-py3.6.egg/taskgraph/Task.py", line 1085, in _call
2021-09-28T23:18:24.6518445Z     payload = result.get()
2021-09-28T23:18:24.6519485Z   File "/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/multiprocessing/pool.py", line 644, in get
2021-09-28T23:18:24.6520345Z     raise self._value
2021-09-28T23:18:24.6520865Z ZeroDivisionError: division by zero
2021-09-28T23:18:24.6521577Z DEBUG    taskgraph.Task:Task.py:816 Invoking terminate. already terminated? False
2021-09-28T23:18:24.6522319Z DEBUG    taskgraph.Task:Task.py:826 shutting down workers
2021-09-28T23:18:24.6522933Z DEBUG    taskgraph.Task:Task.py:811 taskgraph closed
2021-09-28T23:18:24.6523555Z DEBUG    taskgraph.Task:Task.py:769 joining taskgraph
2021-09-28T23:18:24.6524095Z =========================== short test summary info ============================
2021-09-28T23:18:24.6525441Z FAILED tests/test_task.py::TaskGraphTests::test_broken_task - AssertionError:...
2021-09-28T23:18:24.6526113Z ========================= 1 failed, 45 passed in 6.63s =========================
2021-09-28T23:18:24.6994813Z ##[error]Process completed with exit code 1.
```

The failures were always on linux, but sometimes on python 3.7, sometimes on python 3.6.  They were appearing more consistently when using `math.isclose` to compare mtimes (like from `os.path.getmtime()`), and less consistently now that the solution swiched to `os.stat().st_mtime_ns`.

In looking at the test, the important thing is that the test doesn't hang, not that there isn't a `ZeroDivisionError` ... so maybe the test just needs to be adjusted to account for the thing actually being tested.

--- a/HISTORY.rst
+++ b/HISTORY.rst
@@ -6,6 +6,10 @@ TaskGraph Release History
 
 Unreleased Changes
 ------------------
+* Fixed an issue where exceptions raised during execution where the task
+  completed before ``TaskGraph.join()`` was called would not be raised.  Now,
+  if a task raises an exception, its exception will always be raised when
+  both ``Task.join()`` and ``TaskGraph.join()`` is called.
 * Fixed an issue where tasks with ``hash_algorithm='sizetimestamp'`` would,
   under certain conditions, fail to re-execute when they should.  This only
   occurred when a graph writing the same amount of , but possibly different,

--- a/taskgraph/Task.py
+++ b/taskgraph/Task.py
@@ -767,18 +767,13 @@ class TaskGraph(object):
         """
         LOGGER.debug("joining taskgraph")
         if self._n_workers < 0:
+            # Join() is meaningless since tasks execute synchronously.
             LOGGER.debug(
                 'n_workers: %s; join is vacuously true' % self._n_workers)
             return True
 
         if self._terminated:
-            LOGGER.debug('Graph was terminated; checking for exceptions')
-            for task_hash, task in self._task_hash_map.items():
-                if task.exception_object:
-                    LOGGER.debug(
-                        'Task %s had an exception; raising %s' % (
-                            task_hash, task.exception_object))
-                    raise task.exception_object
+            LOGGER.debug('Graph already terminated')
             return True
 
         try:

--- a/taskgraph/Task.py
+++ b/taskgraph/Task.py
@@ -766,11 +766,13 @@ class TaskGraph(object):
 
         """
         LOGGER.debug("joining taskgraph")
-        if self._n_workers < 0 or self._terminated:
+        if self._n_workers < 0:
             LOGGER.debug(
-                'workers: %s; terminated? %s' % (
-                    self._n_workers, self._terminated))
+                'n_workers: %s; join is vacuously true' % self._n_workers)
+            return True
 
+        if self._terminated:
+            LOGGER.debug('Graph was terminated; checking for exceptions')
             for task_hash, task in self._task_hash_map.items():
                 if task.exception_object:
                     LOGGER.debug(
@@ -778,6 +780,7 @@ class TaskGraph(object):
                             task_hash, task.exception_object))
                     raise task.exception_object
             return True
+
         try:
             LOGGER.debug("attempting to join threads")
             timedout = False

--- a/taskgraph/Task.py
+++ b/taskgraph/Task.py
@@ -772,10 +772,6 @@ class TaskGraph(object):
                 'n_workers: %s; join is vacuously true' % self._n_workers)
             return True
 
-        if self._terminated:
-            LOGGER.debug('Graph already terminated')
-            return True
-
         try:
             LOGGER.debug("attempting to join threads")
             timedout = False
@@ -790,7 +786,7 @@ class TaskGraph(object):
                         "task %s timed out in graph join", task.task_name)
                     return False
             if self._closed:
-                # Close down the taskgraph
+                # Close down the taskgraph; ok if already terminated
                 self._executor_ready_event.set()
                 self._terminate()
             return True

--- a/taskgraph/Task.py
+++ b/taskgraph/Task.py
@@ -770,6 +770,13 @@ class TaskGraph(object):
             LOGGER.debug(
                 'workers: %s; terminated? %s' % (
                     self._n_workers, self._terminated))
+
+            for task_hash, task in self._task_hash_map.items():
+                if task.exception_object:
+                    LOGGER.debug(
+                        'Task %s had an exception; raising %s' % (
+                            task_hash, task.exception_object))
+                    raise task.exception_object
             return True
         try:
             LOGGER.debug("attempting to join threads")

--- a/taskgraph/Task.py
+++ b/taskgraph/Task.py
@@ -777,6 +777,8 @@ class TaskGraph(object):
             timedout = False
             for task in self._task_hash_map.values():
                 LOGGER.debug("attempting to join task %s", task.task_name)
+                # task.join() will raise any exception that resulted from the
+                # task's execution.
                 timedout = not task.join(timeout)
                 LOGGER.debug("task %s was joined", task.task_name)
                 # if the last task timed out then we want to timeout for all

--- a/taskgraph/Task.py
+++ b/taskgraph/Task.py
@@ -5,7 +5,6 @@ import hashlib
 import inspect
 import logging
 import logging.handlers
-import math
 import multiprocessing
 import multiprocessing.pool
 import os
@@ -768,6 +767,9 @@ class TaskGraph(object):
         """
         LOGGER.debug("joining taskgraph")
         if self._n_workers < 0 or self._terminated:
+            LOGGER.debug(
+                'workers: %s; terminated? %s' % (
+                    self._n_workers, self._terminated))
             return True
         try:
             LOGGER.debug("attempting to join threads")

--- a/tests/test_task.py
+++ b/tests/test_task.py
@@ -91,12 +91,6 @@ def _div_by_zero():
     return 1/0
 
 
-def _div_by_zero_with_sleep():
-    """Divide by zero to raise an exception."""
-    time.sleep(1)
-    return 1/0
-
-
 def _create_file(target_path, content):
     """Create a file with contents."""
     with open(target_path, 'w') as target_file:
@@ -480,7 +474,7 @@ class TaskGraphTests(unittest.TestCase):
         task_graph = taskgraph.TaskGraph(self.workspace_dir, 1)
 
         _ = task_graph.add_task(
-            func=_div_by_zero_with_sleep, task_name='test_broken_task')
+            func=_div_by_zero, task_name='test_broken_task')
         task_graph.close()
         with self.assertRaises(ZeroDivisionError):
             task_graph.join()

--- a/tests/test_task.py
+++ b/tests/test_task.py
@@ -1254,7 +1254,8 @@ class TaskGraphTests(unittest.TestCase):
                         transient_run=True,
                         store_result=True,
                         args=(expected_value,),
-                        task_name='expected error {iteration_id}')
+                        task_name=f'expected error {iteration_id}')
+
                     value = value_task.get()
 
                 with self.assertRaises(RuntimeError):

--- a/tests/test_task.py
+++ b/tests/test_task.py
@@ -4,8 +4,8 @@ import logging
 import logging.handlers
 import multiprocessing
 import os
-import pickle
 import pathlib
+import pickle
 import re
 import rstcheck
 import shutil
@@ -88,6 +88,12 @@ def _sum_lists_from_disk(list_a_path, list_b_path, target_path):
 
 def _div_by_zero():
     """Divide by zero to raise an exception."""
+    return 1/0
+
+
+def _div_by_zero_with_sleep():
+    """Divide by zero to raise an exception."""
+    time.sleep(1)
     return 1/0
 
 
@@ -472,8 +478,9 @@ class TaskGraphTests(unittest.TestCase):
     def test_broken_task(self):
         """TaskGraph: Test that a task with an exception won't hang."""
         task_graph = taskgraph.TaskGraph(self.workspace_dir, 1)
+
         _ = task_graph.add_task(
-            func=_div_by_zero, task_name='test_broken_task')
+            func=_div_by_zero_with_sleep, task_name='test_broken_task')
         task_graph.close()
         with self.assertRaises(ZeroDivisionError):
             task_graph.join()

--- a/tests/test_task.py
+++ b/tests/test_task.py
@@ -479,7 +479,9 @@ class TaskGraphTests(unittest.TestCase):
             _ = broken_task.join()
 
         task_graph.close()
-        task_graph.join()
+
+        with self.assertRaises(ZeroDivisionError):
+            task_graph.join()
 
     def test_broken_task_chain(self):
         """TaskGraph: test dependent tasks fail on ancestor fail."""
@@ -1243,6 +1245,8 @@ class TaskGraphTests(unittest.TestCase):
                     task_name='first re-run transient')
                 value = value_task.get()
                 self.assertEqual(value, expected_value)
+                task_graph.close()
+                task_graph.join()
             else:
                 with self.assertRaises(RuntimeError):
                     value_task = task_graph.add_task(
@@ -1253,8 +1257,9 @@ class TaskGraphTests(unittest.TestCase):
                         task_name='expected error {iteration_id}')
                     value = value_task.get()
 
-            task_graph.close()
-            task_graph.join()
+                with self.assertRaises(RuntimeError):
+                    task_graph.join()
+
             task_graph = None
 
     def test_malformed_taskgraph_database(self):

--- a/tests/test_task.py
+++ b/tests/test_task.py
@@ -473,11 +473,13 @@ class TaskGraphTests(unittest.TestCase):
         """TaskGraph: Test that a task with an exception won't hang."""
         task_graph = taskgraph.TaskGraph(self.workspace_dir, 1)
 
-        _ = task_graph.add_task(
+        broken_task = task_graph.add_task(
             func=_div_by_zero, task_name='test_broken_task')
-        task_graph.close()
         with self.assertRaises(ZeroDivisionError):
-            task_graph.join()
+            _ = broken_task.join()
+
+        task_graph.close()
+        task_graph.join()
 
     def test_broken_task_chain(self):
         """TaskGraph: test dependent tasks fail on ancestor fail."""
@@ -1190,7 +1192,6 @@ class TaskGraphTests(unittest.TestCase):
         with open(target_file_path, 'r') as target_file:
             self.assertEqual(target_file.read(), 'content')
 
-
     def test_return_value_no_record(self):
         """TaskGraph: test  ``get`` raises exception if not set to record."""
         task_graph = taskgraph.TaskGraph(self.workspace_dir, -1)
@@ -1239,7 +1240,7 @@ class TaskGraphTests(unittest.TestCase):
                     transient_run=True,
                     store_result=True,
                     args=(expected_value,),
-                    task_name=f'first re-run transient')
+                    task_name='first re-run transient')
                 value = value_task.get()
                 self.assertEqual(value, expected_value)
             else:
@@ -1249,7 +1250,7 @@ class TaskGraphTests(unittest.TestCase):
                         transient_run=True,
                         store_result=True,
                         args=(expected_value,),
-                        task_name=f'expected error {iteration_id}')
+                        task_name='expected error {iteration_id}')
                     value = value_task.get()
 
             task_graph.close()
@@ -1301,7 +1302,7 @@ class TaskGraphTests(unittest.TestCase):
                 'task_reexecution_hash', 'target_path_stats', 'result']
             connection = sqlite3.connect(database_path)
             cursor = connection.cursor()
-            cursor.execute(f'PRAGMA table_info(taskgraph_data)')
+            cursor.execute('PRAGMA table_info(taskgraph_data)')
             result = list(cursor.fetchall())
             cursor.close()
             connection.commit()

